sorc/build_all.sh

```diff
@@ -105,170 +107,190 @@ ERRSCRIPT=${ERRSCRIPT:-'eval [[ $err = 0 ]]'}
-    fi
-    err=$((err + rc))
-  fi
-else
-  echo " .... Skip building GDASApp  .... "
-fi
+echo "Building ${build_list}"
 
-#------------------------------------
-# build gsi monitor
-#------------------------------------
-if [[ -d gsi_monitor.fd ]]; then
-  if [[ ${Build_gsi_monitor} == 'true' ]]; then
-    echo " .... Building gsi monitor .... "
-    # shellcheck disable=SC2086,SC2248
-    ./build_gsi_monitor.sh ${_ops_opt} ${_verbose_opt} > "${logs_dir}/build_gsi_monitor.log" 2>&1
-    # shellcheck disable=
-    rc=$?
-    if (( rc != 0 )) ; then
-      echo "Fatal error in building gsi monitor."
-      echo "The log file is in ${logs_dir}/build_gsi_monitor.log"
-    fi
-    err=$((err + rc))
-  fi
-else
-  echo " .... Skip building gsi monitor .... "
+# Go through all builds and adjust CPU counts up if possible
+if [[ ${requested_cpus} -lt ${_build_job_max} && ${big_jobs} -gt 0 ]]; then
+   # Add cores to the gdas, ufs, and gsi build jobs
+   extra_cores=$(( _build_job_max - requested_cpus ))
+   extra_cores=$(( extra_cores / big_jobs ))
+   for build in "${!build_jobs[@]}"; do
+      if [[ "${build}" == "gdas" || "${build}" == "ufs_model" || "${build}" == "gsi_enkf" ]]; then
+         build_jobs[${build}]=$(( build_jobs[${build}] + extra_cores ))
+      fi
+   done
 fi
 
-#------------------------------------
-# build UPP
-#------------------------------------
-if [[ ${Build_upp} == 'true' ]]; then
-  echo " .... Building UPP .... "
-  # shellcheck disable=SC2086,SC2248
-  ./build_upp.sh ${_ops_opt} ${_verbose_opt} > "${logs_dir}/build_upp.log" 2>&1
-  # shellcheck disable=
-  rc=$?
-  if (( rc != 0 )) ; then
-    echo "Fatal error in building UPP."
-    echo "The log file is in ${logs_dir}/build_upp.log"
-  fi
-  err=$((err + rc))
-fi
+procs_in_use=0
+declare -A build_ids
 
-#------------------------------------
-# build ufs_utils
-#------------------------------------
-if [[ ${Build_ufs_utils} == 'true' ]]; then
-  echo " .... Building ufs_utils .... "
-  # shellcheck disable=SC2086,SC2248
-  ./build_ufs_utils.sh ${_verbose_opt} > "${logs_dir}/build_ufs_utils.log" 2>&1
-  # shellcheck disable=
-  rc=$?
-  if (( rc != 0 )) ; then
-    echo "Fatal error in building ufs_utils."
-    echo "The log file is in ${logs_dir}/build_ufs_utils.log"
-  fi
-  err=$((err + rc))
-fi
+builds_started=0
+# Now start looping through all of the remaining jobs until everything is done
+while [[ ${builds_started} -lt ${#build_jobs[@]} ]]; do
+   if [[ ${procs_in_use} -lt ${_build_job_max} ]]; then
+      if [[ -n "${build_jobs[gdas]+0}" && -z "${build_ids[gdas]+0}" ]]; then
+         if [[ ${_build_job_max} -ge $(( build_jobs['gdas'] + procs_in_use )) ]]; then
+            ./build_gdas.sh -j "${build_jobs[gdas]}" "${_verbose_opt}" > \
+               "${logs_dir}/build_gdas.log" 2>&1 &
+            build_ids["gdas"]=$!
+            echo "Starting build_gdas.sh"
+            procs_in_use=$(( procs_in_use + build_jobs['gdas'] ))
+         fi
+      fi
```

Comment:
This bit of code is getting repeated over and over. Can we turn this into a loop?

Reply:
Probably, but just noting that some of the build scripts take different arguments than others.  For instance, build_upp.sh does not take a `-j` and build_ufs.sh takes a `-a` argument for the `$APP` of choice.  I could create an additional array that stores the build command for each application, which would be fewer lines overall and make it easier for developers to find/adjust those build commands as needed.  I will give that a try.

Reply:
I made an implementation which seemed to be working, but it has revealed a case where the WW3 pre/post applications attempt to build before the UFS, which causes an issue as those require some modules that are only compiled when the entire UFS is compiled.  I will need to add another if block to ensure `build_ww3prepost.sh` doesn't execute until `build_ufs.sh` completes.  This also means I can throw more CPUs at WW3 since it won't be competing with the UFS.

Reply:
@DavidHuber-NOAA  This is a bit confusing to me.  Depending on where things are built I could see coming both UFS and WW3 pre/post at the same time would potentially be an issue but perhaps this just needs a script change.  Also FYI - I have an update coming that will add an extra build option to both UFS and WW3 for unstructured grid - I'm sure it'll be easy to implement the update no matter which goes first, just wanted you to be aware.

Reply:
@JessicaMeixner-NOAA Hmm, your suspicion is correct, but this leaves me a bit perplexed.   The last time I ran the build_all.sh script, buid_ww3prepost.sh launched before build_ufs.sh and crashed with several undefined reference errors.  For instance:
```
[ 72%] Linking Fortran executable ../../bin/ww3_strt
../../lib/libww3.a(w3gdatmd.F90.o): In function `w3gdatmd_mp_w3gntx_':
/scratch1/NCEPDEV/nems/David.Huber/GW/gw_parallel_builds/sorc/ufs_model.fd/WW3/model/src/w3gdatmd.F90:2953: undefined reference to `w3gsrumd_mp_w3gsuc_ptr_r8_'
/scratch1/NCEPDEV/nems/David.Huber/GW/gw_parallel_builds/sorc/ufs_model.fd/WW3/model/src/w3gdatmd.F90:3001: undefined reference to `w3gsrumd_mp_w3cgdm_r4_'
/scratch1/NCEPDEV/nems/David.Huber/GW/gw_parallel_builds/sorc/ufs_model.fd/WW3/model/src/w3gdatmd.F90:2986: undefined reference to `w3gsrumd_mp_w3cgdm_r4_'
```
However, when I review an older build log, I see that the missing module, `w3gsrumd`, was built during the ww3prepost job.  I'm not sure how this could happen.  Any ideas?  I will save the log file here if you wish to take a look while I continue testing: `/scratch1/NCEPDEV/nems/David.Huber/GW/gw_parallel_builds/sorc/logs/build_ww3prepost.save.log`.

Reply:
@DavidHuber-NOAA I've built using the WW3 script without ever building UFS, I'll take a look at your log file.

Reply:
@DavidHuber-NOAA so the WW3 pre/post builds multiple executables.  Is this an issue that you are trying to build at the top level with multiple processors instead of the actual cmake command having multiple processors?  (I ask because something similar happened a long time ago).  So for example, the top level build_ww3prepost should not be called with multiple MPI processors but within it the make commands could use multiple?    If that doesn't make sense let me know and I can try to explain in person...

Reply:
I have implemented a new array that keeps track of build options and consolidated the build if blocks.